File: unvisited_stones/unvisited_stones.php
<?php
/**
 * Находит количество валунов, которые не посетит ни одна птица.
 * 
 * @param int $N Количество птиц.
 * @param int $stones Общее количество валунов.
 * @param array $birds Массив с дистанциями перелёта каждой птицы.
 * @return int Количество непосещённых валунов.
 */

function unvisitedStones($N, $stones, $birds) {
    $visited = array_fill(0, $stones + 1, false);
    foreach ($birds as $bird) {
        for ($pos = $bird; $pos <= $stones; $pos += $bird) {
            $visited[$pos] = true;
        }
    }
    return count(array_filter(array_slice($visited, 1), function($v) { return !$v; }));
}
?>

File: unvisited_stones/UnvisitedStones.kt
/**
 * Находит количество валунов, которые не посетит ни одна птица.
 * 
 * @param N Количество птиц.
 * @param stones Общее количество валунов.
 * @param birds Массив с дистанциями перелёта каждой птицы.
 * @return Количество непосещённых валунов.
 */
 
fun unvisitedStones(N: Int, stones: Int, birds: IntArray): Int {
    val visited = BooleanArray(stones + 1) // Индексация с 1
    for (bird in birds) {
        var pos = bird
        while (pos <= stones) {
            visited[pos] = true
            pos += bird
        }
    }
    return visited.count { !it } - 1 // Игнорируем 0-й индекс
}